### PR TITLE
Fix errors: conrtib/build_{executable,sysimg}.jl

### PR DESCRIPTION
**Pull request by KDr2**
- fix few compile error
- do _NOT_ build both release file and debug file at once, build a release file by default, build a debug version by `--debug` option
- add option `--portable` to build a portable executable file which can be run with LD_LIBRARY_PATH and JULIA_SYSIMG environment var set.

### CODE REVIEW

**Comment by ViralBShah** [comment]
Cc: @nalimilan

File: contrib/build_executable.jl [hunk 1]
```diff
@@ -123,47 +127,48 @@ function build_executable(exename, script_file, targetdir=nothing, cpu_target="n
...
+    linker_rpath = portable ? `` : `-Wl,-rpath,$(sys.buildpath)`
+    println("running: $gcc -g -Wl,--no-as-needed $win_arg $(join(incs, " ")) $(cfile) -o $(exe_file.buildfile) $(linker_rpath) -L$(sys.buildpath) $(exe_file.libjulia) -l$(exename)")
+    cmd = setenv(`$gcc -g -Wl,--no-as-needed $win_arg $(incs) $(cfile) -o $(exe_file.buildfile) $(linker_rpath) -L$(sys.buildpath) $(exe_file.libjulia) -l$(exename)`, ENV2)
+    run(cmd)
+    println()
 
-    println("running: rm -rf $(tmpdir) $(sys.buildfile).o $(sys.buildfile0).o $(sys.buildfile0).ji")
-    map(f-> rm(f, recursive=true), [tmpdir, sys.buildfile*".o", sys.buildfile0*".o", sys.buildfile0*".ji"])
+    println("running: rm -rf $(tmpdir) $(sys.buildfile * ".o") $(sys.buildfile0 * ".o") $(sys.buildfile0 * ".ji")")
+    run(`rm -rf $(tmpdir) $(sys.buildfile * ".o") $(sys.buildfile0 * ".o") $(sys.buildfile0 * ".ji")`)
```
Comment (tkelman) [inline contrib/build_executable.jl]: use the julia `rm` function, not the coreutils `rm` which is not going to be present on windows

Comment (KDr2) [inline contrib/build_executable.jl]: done, thanks.

File: contrib/build_executable.jl [hunk 2]
```diff
@@ -46,16 +46,21 @@ type SysFile
...
     SysFile(buildpath, buildfile, buildfile0)
 end
 
-function build_executable(exename, script_file, targetdir=nothing, cpu_target="native"; force=false)
-    julia = abspath(joinpath(JULIA_HOME, "julia"))
+function build_executable(exename, script_file, targetdir=nothing, cpu_target="native";
+                          force=false, debug=false, portable=false)
+    julia = abspath(joinpath(JULIA_HOME, debug ? "julia-debug" : "julia"))
+    if !isfile(julia)
+        println("ERROR: file `$(julia)' not found.")
```
Comment (nalimilan) [inline contrib/build_executable.jl]: The first quote isn't straight here.

File: contrib/build_executable.jl [hunk 3]
```diff
@@ -123,47 +127,52 @@ function build_executable(exename, script_file, targetdir=nothing, cpu_target="n
...
             end
         end
     end
 
-    for exe in [exe_release, exe_debug]
-        println("$(exe.targetfile) successfully created.")
+    println("$(exe_file.targetfile) successfully created.")
+    if portable
+        println("The executable file is portable," *
+                "please set the library search path(LD_LIBRARY_PATH) and JULIA_SYSIMG environment variable before run it.")
```
Comment (nalimilan) [inline contrib/build_executable.jl]: "before running it"

Comment (nalimilan) [inline contrib/build_executable.jl]: ...and a space before "(LD_LIBRARY_PATH)". Also, "and the JULIA_SYSIMG".

**Comment by nalimilan** [comment]
@ViralBShah Did you mean to CC somebody else? I've no particular knowledge about this file.

File: contrib/build_executable.jl (hunk 3 above)
Comment (KDr2) [inline contrib/build_executable.jl]: fixed, :blush:

**Comment by tkelman** [comment]
@dhoegh has been pretty actively maintaining a version of this at https://github.com/dhoegh/BuildExecutable.jl - it might be worth actually migrating so there aren't 2 separate copies of this going forward

**Comment by dhoegh** [comment]
Yes I would deferentially appreciate the help at https://github.com/dhoegh/BuildExecutable.jl . I have already fixed some of the issues mentioned here. I support the idea of removing `build_executable.jl` from base as it is untested in base and the code is self-sustained. It is easier to fix bugs in the script, as `BuildExecutable.jl` do not need to follow base releases. At `BuildExecutable.jl` the code is tested on both Travis and Appveyor.

**Comment by KDr2** [comment]
@dhoegh great work.
I would like to migrate the code to @dhoegh 's BuildExecutable.jl. But I should do this in my spare time, it may take more or less a week, I'll delete these two files from the julia base repo after that. This PR can be merged and I'll create a new one for the migration/deleting, Or just leave it here util the migration has been done.  Is this a working plan?  @tkelman @nalimilan @ViralBShah @dhoegh

**Comment by dhoegh** [comment]
My opinion is that `build_sysimg.jl` belongs in base and only `build_executable.jl` should be migrated and removed from base. I do only carry the `build_sysimg.jl` in the repository to fix `build_executable.jl` on 0.4, see  #14018. When 0.4.2 is released I will bump the requirement to 0.4.2 and remove `build_sysimg.jl` from the repository.

**Comment by tkelman** [comment]
That makes sense to me. sysimg is needed if you want to make a change to base source and actually have it take effect, which can be a bit counterintuitive for newcomers.

**Comment by KDr2** [comment]
I think this PR is ready to merge, I've migrated the change to @dhoegh's  https://github.com/dhoegh/BuildExecutable.jl at https://github.com/dhoegh/BuildExecutable.jl/pull/6 .

**Comment by tkelman** [comment]
Note the travis errors, `make install` needs to be updated for the removal of `build_executable.jl`. Do a `git grep -n build_executable.jl` to find if it's mentioned anywhere else.

**Comment by KDr2** [comment]
Done, thanks. @tkelman

**Comment by tkelman** [comment]
Travis failure was an unrelated repl test issue caused by https://github.com/JuliaLang/julia/commit/fb48014b23273bc3f021d31aab3246d717d2bb5d. Thanks for the contribution!

**Comment by dhoegh** [comment]
Could the changes to `build_sysimg.jl` be backported to 0.4?

**Comment by tkelman** [comment]
I guess? The debug flag is a bit feature-y.

**Comment by dhoegh** [comment]
I see. If this got backported I would be able to get rid of the local copy of `build_sysimg.jl` I carry in https://github.com/dhoegh/BuildExecutable.jl which I would be nice.

**Comment by tkelman** [comment]
I thought the osx fix that did get backported was enough for that? What does having the debug flag fix?

**Comment by dhoegh** [comment]
In https://github.com/dhoegh/BuildExecutable.jl/pull/6 @KDr2 states that
> If I only have julia or julia-debug in my JULIA_HOME, it fails to build.
I do not know how @KDr2 is building/using julia but apparently in a way, where julia and julia-debug is not contained within JULIA_HOME.

**Comment by tkelman** [comment]
In a source build you might only have release but not debug present, or maybe the other way around.

I'm a bit skeptical of some od the changes over there with respect to JULIA_HOME, you should not be setting that environment variable or modifying it.

**Comment by KDr2** [comment]
* About the `debug` flag: Yeah, I usually build julia from source with `make`, or `make julia-debug` occasionally, and sometime I do `make clean(all)?`, so `julia` and `julia-debug` may not exist at the same time.

* About unsetting/setting `JULIA_HOME`: I found `jl_init_with_image` we are using calls `julia_init` with `JL_IMAGE_JULIA_HOME`, when `JULIA_HOME` is set(usually to the bin directory under the julia installation), it will lookup the system image for our built executable  under `JULIA_HOME`, not the directory where the executable placed. so I unset the `JULIA_HOME` before calling `jl_init_with_image` and reset it after the call. Maybe there's another better solution?

@tkelman @dhoegh

**Comment by tkelman** [comment]
Don't set `JULIA_HOME` in the first place?

**Comment by KDr2** [comment]
Yes, don't set `JULIA_HOME` in the first place is the right way :smile: , and the c code will do nothing if we have no `JULIA_HOME` set.

But sometimes the `JULIA_HOME` is set
- by accident(I mean not by user's intent, just forget to unset it, or set by a program which exec our executable...)
- because the julia program which we are packaging is using `ENV["JULIA_HOME"]` for some purpose.


Or we can simply and clearly write this down into the document.

**Comment by tkelman** [comment]
That environment variable is significant to the julia runtime, so it shouldn't be used for arbitrary other purposes. That significance should be documented, or possibly removed if the original reasons for it are no longer applicable.

**Pull request merged**